File: branches/wire/remote/server/src/java/org/openqa/selenium/remote/server/handler/ExecuteScript.java
/*
Copyright 2007-2009 WebDriver committers
Copyright 2007-2009 Google Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package org.openqa.selenium.remote.server.handler;

import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.remote.Response;
import org.openqa.selenium.remote.server.DriverSessions;
import org.openqa.selenium.remote.server.JsonParametersAware;
import org.openqa.selenium.remote.server.KnownElements;
import org.openqa.selenium.remote.server.rest.ResultType;

import com.google.common.base.Function;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ExecuteScript extends WebDriverHandler implements JsonParametersAware {
  private Response response;
  private String script;
  private List<Object> args = new ArrayList<Object>();

  public ExecuteScript(DriverSessions sessions) {
    super(sessions);
  }

  public void setJsonParameters(Map<String, Object> allParameters) throws Exception {
    script = (String) allParameters.get("script");

    List<?> params = (List<?>) allParameters.get("args");

    parseParams(params, args);
  }
  
  private void parseParams(List<?> params, List<Object> args) {
    for (Object param : params) {
      if (param instanceof Map) {
        Map<String, Object> paramAsMap = (Map<String, Object>)param;
        String type = (String) paramAsMap.get("type");
        if ("ELEMENT".equals(type)) {
          KnownElements.ProxiedElement element = (KnownElements.ProxiedElement) getKnownElements().get((String) paramAsMap.get("value"));
          args.add(element.getWrappedElement());
        } else {
          args.add(paramAsMap.get("value"));
        }
      } else if (param instanceof List) {
        List<Object> sublist = new ArrayList<Object>();
        parseParams((List<?>)param, sublist);
        args.add(sublist);
      }
    }
  }

  public ResultType call() throws Exception {
    response = newResponse();

    Object value;
    if (args.size() > 0) {
      value = ((JavascriptExecutor) getDriver()).executeScript(script, args.toArray());
    } else {
      value = ((JavascriptExecutor) getDriver()).executeScript(script);
    }

    Object result = new ResultConverter().apply(value);
    response.setValue(result);
    
    return ResultType.SUCCESS;
  }

  public Response getResponse() {
    return response;
  }
  
  @Override
  public String toString() {
    return String.format("[execute script: %s, %s]", script, args);
  }

  /**
   * Converts an object to be sent as JSON according to the wire protocol.
   */
  private class ResultConverter implements Function<Object, Object> {
    public Object apply(Object result) {
      if (result instanceof WebElement) {
        String elementId = getKnownElements().add((WebElement) result);
        return ImmutableMap.of("ELEMENT", elementId);
      }

      if (result instanceof List) {
        @SuppressWarnings("unchecked")
        List<Object> resultAsList = (List<Object>) result;
        return Lists.newArrayList(Iterables.transform(resultAsList, this));
      }

      return result;
    }
  }
}

File: tags/selenium-2.12.0/javascript/webdriver-atoms/inject/execute_script.js
// Copyright 2011 WebDriver committers
// Copyright 2011 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * @fileoverview Wrapping execute script to use a serialized window object.
 *
 */

goog.provide('webdriver.inject');

goog.require('bot.inject');
goog.require('bot.inject.cache');

/**
 * Wrapper to allow passing a seliazed window object to executeScript.
 *
 * @param {!(string|function)} fn The function to execute.
 * @param {Array.<*>} args Array of arguments to pass to fn.
 * @param {!{bot.inject.WINDOW_KEY:string}=} opt_window The serialized window
 *     object to be read from the cache.
 * @return {!(string|bot.inject.Response)} The response object. If
 *     opt_stringify is true, the result will be serialized and returned in
 *     string format.
 */
webdriver.inject.executeScript = function(fn, args, opt_window) {
  return bot.inject.executeScript(
      fn, args, true, webdriver.inject.getWindow_(opt_window));
};

/**
 *
 * @param {!(string|function)} fn The function to execute.
 * @param {Array.<*>} args Array of arguments to pass to fn.
 * @param {int} timeout The timeout to wait up to in millis.
 * @param {!{bot.inject.WINDOW_KEY:string}=} opt_window The serialized window
 * object to be read from the cache.
 * @return {!(string|bot.inject.Response)} The response object. If
 *     opt_stringify is true, the result will be serialized and returned in
 *     string format.
 */
webdriver.inject.executeAsyncScript =
    function(fn, args, timeout, onDone, opt_window) {
  return bot.inject.executeScript(fn, args, timeout, onDone, true,
          webdriver.inject.getWindow_(opt_window));
};


/**
 * Get the window to use.
 *
 * @param {!Window=} opt_window An optional window to use.
 * @return {!Window} A reference to a window.
 * @private
 */
webdriver.inject.getWindow_ = function(opt_window) {
  var win;
  if (opt_window) {
    win = bot.inject.cache.getElement(opt_window['WINDOW']);
  } else {
    win = window;
  }
  return win;
};
